### PR TITLE
[WIP] draft idea for client generators

### PR DESCRIPTION
**Pull request by stevekuznetsov**

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
We should probably determine what we want users to be able to do. Today, they can:

 - pass around a factory (`kubernetes.Interface`, `rbacv1.Interface`) that can:
   - create a cluster-scoped client for cluster-scoped resources (`kubernetes.Interface.Core().V1().Nodes()`)
   - create a cluster-scoped (namespace-deferred) client for namespaced resources (`kubernetes.Interface.Core().V1().(*corev1.ConfigMapsGetter)`), which can:
     - create a cluster-scoped, namespace-scoped client for namespaced resources (`kubernetes.Interface.Core().V1().ConfigMaps(namespace)`)
   
With our generated helpers, we could aim to keep those:
- pass around a cluster-less factory (`kcpkubernetes.Interface`) that can:
  - create a cluster-deferred client for cluster-scoped resources (`kcpkubernetes.Interface.Core().V1().Nodes()`), which can
    - create a cluster-scoped client for cluster-scoped resources (`kcpkubernetes.Interface.Core().V1().Nodes().Cluster(name)`)
  - create a cluster-deferred, namespace-deferred client for namespaced resources (`kcpkubernetes.Interface.Core().V1().ConfigMaps()`), which can:
    - create a cluster-scoped, namespace-deferred client for namespaced resources (`kcpkubernetes.Interface.Core().V1().ConfigMaps().Cluster(name)`), which can:
      - create a cluster-scoped, namespace-scoped client for namespaced resources (`kcpkubernetes.Interface.Core().V1().ConfigMaps().Cluster(name).Namespace(namespace)`)
    - create a cluster-deferred, namespace-scoped client for namespaced resources (`kcpkubernetes.Interface.Core().V1().ConfigMaps().Namespace(namespace)`), which can:
      - create a cluster-scoped, namespace-scoped client for namespaced resources (`kcpkubernetes.Interface.Core().V1().ConfigMaps().Namespace(namespace).Cluster(name)`)

**Comment by stevekuznetsov** [comment]
If we wanted to remove one level of nesting there we could e.g.

```go
type NodeClusterGetter interface {
    Nodes(logicalcluster.Name) NodeInterface
}

type ConfigMapClusterGetter interface {
    ConfigMaps(logicalcluster.Name) ConfigMapInterface
}

type ConfigMapClusterInterface interface {
    Namespace(string) corev1.ConfigMapInterface
}

type ConfigMapNamespaceGetter interface {
    ConfigMapsByNamespace(string) ConfigMapClusterScoper
}

type ConfigMapClusterScoper interface {
    Cluster(logicalcluster.Name) corev1.ConfigMapInterface
}

func main() {
    var clientFactory *kcpkubernetes.Interface
    unscopedNodeClient := clientFactory.Core().V1().(NodeClusterGetter)

    // this feels normal
    unscopedConfigMapClusterClient := clientFactory.Core().V1().(ConfigMapGetter)
    unnamespacedConfigMapClient := unscopedConfigMapClusterClient.ConfigMaps(logicalcluster.New("whoa"))
    namespacedConfigMapClient := unnamespacedConfigMapClient.Namespace("foo")
    
    // this feels weird
    unscopedConfigMapNamespaceClient := clientFactory.Core().V1().(ConfigMapNamespaceGetter)
    unclusteredConfigMapClient := unscopedConfigMapNamespaceClient.ConfigMapsByNamespace("foo")
    fullyQualifiedConfigMapClient := unclusteredConfigMapClient.Cluster(logicalcluster.New("whoa"))
}
```

File: main.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	// now, use it as normal
+	rbacClient.RoleBindings("default").Get(context.TODO(), "system", metav1.GetOptions{})
+
+	// defer scoping until later, keeping un-scoped but strongly typed clients for specific resources
+	clusterRoleClient := unscopedClient.ClusterRoles()
+	roleClient := unscopedClient.Roles()
+	// use them *almost* as normal
+	clusterRoleClient.Cluster(logicalcluster.New("whoa")).ClusterRoles().Get(context.TODO(), "foo", metav1.GetOptions{})
+	// NOTE: for cluster-scoped resources we could technically make the above call not stutter on e.g. ClusterRoles():
+	// clusterRoleClient.Cluster(logicalcluster.New("whoa")).Get(context.TODO(), "foo", metav1.GetOptions{})
```
Comment (p0lyn0mial) [inline main.go:31]: yes, I think this looks more natural.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+	roleClient.Cluster(logicalcluster.New("dang")).Roles("default").Get(context.TODO(), "bar", metav1.GetOptions{})
+	// NOTE: for namespace-scoped resources we could technically make the above call not stutter on e.g. Roles():
+	// roleClient.Cluster(logicalcluster.New("dang")).Namespace("default").Get(context.TODO(), "foo", metav1.GetOptions{})
```
Comment (p0lyn0mial) [inline main.go:35]: Although this looks really good this is not how a standard client-go interface looks like, right?

A standard client-go interface exposes `RolesGetter` that allows for `Roles(namespace)` not for `Namespace(namespace)`. It seems we should conform to the standard interface.

File: main.go (hunk 1 above)
Comment (ncdc) [reply main.go:31]: If it's possible to do this, agreed

File: main.go (hunk 2 above)
Comment (ncdc) [reply main.go:35]: If we are _not_ trying to be interface-compatible with existing generated clientsets, I think we should prioritize a DX that "makes sense", i.e. `.Namespace(ns)` >>> `.Roles(ns)`

Comment (ncdc) [reply main.go:35]: Oh that's not what I was suggesting. Huge +1 to cluster, then ns

Comment (stevekuznetsov) [reply main.go:35]: Even if we managed to be interface-compatible for the initial `.ConfigMaps(ns)` call, we'd have to not be as it must be followed by a `.Cluster()` call before using e.g. `.Get()`, so I'm not sure compatibility is critical as it's going to require a refactor anyway.

Comment (ncdc) [reply main.go:35]: Interface compat is only required if you are unwilling or unable to modify existing code. The latter is e.g. if you're using a 3rd party library that uses generated clientsets... we should probably make sure we don't lose sight of this case.

Comment (stevekuznetsov) [reply main.go:35]: Yep, that case I think is captured well in the design doc. To my eyes we should be able to support it by ensuring that we emit the underlying (`client-gen`) types after cluster scoping, no? Then a multi-cluster context can be bolted on top of an existing controller (and they would run one controller per lcluster?)

**Comment by ncdc** [comment]
@stevekuznetsov outdated / close?

**Pull request closed without merging**